waspc/data/Generator/templates/server/src/core/auth/middleware.js

```diff
@@ -0,0 +1,86 @@
+{{={= =}=}}
+import { hashPassword } from '../auth.js'
+import AuthError from '../AuthError.js'
+
+const EMAIL_FIELD = 'email'
+const PASSWORD_FIELD = 'password'
+
+const validateUserData = (data, args, action) => {
+  data = data || {}
+
+  // By default, a given validation v will only run if data.hasOwnProperty(v.validates)
```

Comment:
Ok just had a short discussion with @matijaSos about this, since he also thought (on his own) that this part is hard to understand, which I think we all agree.

I had a short brainstorming with him and we arrived at following solution, which it seems to me solves the problem equally well but is somewhat simpler:
We keep `validates` field, but we remove `always_on`. We modify the validator execution algorithm so that it always runs all the validators on `create`, but for `update` it runs only those for which corresponding fields exist.
This works ok for all of the validators we currently have, but seems to be problematic if there is an optional field in the User entity. However, when we consider it a bit more, we can conclude that users can solve this by putting a conditional in the `validator` function! So, they could do `data => !!data.nickname || data.nickname.length < 32` if field is optional, while if it is required, they would do `data => !!data.nickname && data.nickname.length < 32`.
I am not sure if I explained this correctly or if I missed something, but I am pretty sure this should be ok?

This way we don't have to explain `always_on`, which is pretty hard to explain as a concept!

Reply:
Oof my Ruby and snake case will be a problem for a while 🤦🏻  Yeah I went back and forth on that `alwaysOn` name like you did. Agree on keeping eyes toward extending to multiple fields in future.

And yes!! I love the idea of always running on `create` and optionally on `update`. Easier to explain. I think the check for optional makes sense.

Lastly, the only other thing I was thinking of an enhancement was to have the outer logic of the validator pluck the `validates` field and pass that in only, instead of `data` to the `validator` so you have something like:
```js
{
  validates: 'password',
  message: 'password must be at least 8 characters',
  validator: password => password.length >= 8
},
```

I think it would make it safer and less verbose in many respects. If we choose to make `validates` support multiple we can go back to `data.` access, but for now I think that may tidy it up a bit.

Thanks for the good idea on simplifying it! I wasn't happy with it and this is better 😄

Reply:
This is addressed here: https://github.com/wasp-lang/wasp/pull/376/commits/ba6479e6b0e346a7a75a608243c69de58e1ab908 Thanks!